Test failure about root progress dispatch

This looks like a flaky test, but I don't remember having seen this. It means the progress event dispatched more completion than the total expected. It's concerning, so I'll dig into it.

```
    * What went wrong:
    Execution failed for task ':jibDockerBuild'.
    > Progress exceeds max for 'building image to Docker daemon': 1 more beyond 5

    * Try:
    Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

    * Get more help at https://help.gradle.org

    BUILD FAILED in 4s
    3 actionable tasks: 3 executed
        at org.gradle.testkit.runner.internal.DefaultGradleRunner$1.execute(DefaultGradleRunner.java:237)
        at org.gradle.testkit.runner.internal.DefaultGradleRunner$1.execute(DefaultGradleRunner.java:234)
        at org.gradle.testkit.runner.internal.DefaultGradleRunner.run(DefaultGradleRunner.java:298)
        at org.gradle.testkit.runner.internal.DefaultGradleRunner.build(DefaultGradleRunner.java:234)
        at com.google.cloud.tools.jib.gradle.TestProject.build(TestProject.java:82)
        at com.google.cloud.tools.jib.gradle.JibRunHelper.buildToDockerDaemon(JibRunHelper.java:96)
        at com.google.cloud.tools.jib.gradle.EmptyProjectIntegrationTest.testDockerDaemon_userNames(EmptyProjectIntegrationTest.java:112)
```

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/builder/ProgressEventDispatcher.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/builder/ProgressEventDispatcher.java
@@ -147,7 +147,7 @@ public class ProgressEventDispatcher implements Closeable {
   private long decrementRemainingAllocationUnits(long units) {
     Preconditions.checkState(!closed);
 
-    if (remainingAllocationUnits > units) {
+    if (remainingAllocationUnits >= units) {
       remainingAllocationUnits -= units;
       return units;
     }

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java
@@ -32,9 +32,7 @@ import com.google.common.util.concurrent.Futures;
 import com.google.common.util.concurrent.ListeningExecutorService;
 import com.google.common.util.concurrent.MoreExecutors;
 import java.nio.file.Path;
-import java.util.ArrayDeque;
 import java.util.ArrayList;
-import java.util.Deque;
 import java.util.List;
 import java.util.concurrent.Callable;
 import java.util.concurrent.ExecutionException;
@@ -103,8 +101,7 @@ public class StepsRunner {
   private final List<Runnable> stepsToRun = new ArrayList<>();
 
   @Nullable private String rootProgressDescription;
-  private final Deque<ProgressEventDispatcher.Factory> childProgressDispatcherFactories =
-      new ArrayDeque<>();
+  @Nullable private ProgressEventDispatcher rootProgressDispatcher;
 
   private StepsRunner(
       ListeningExecutorService executorService, BuildConfiguration buildConfiguration) {
@@ -113,66 +110,87 @@ public class StepsRunner {
   }
 
   private void retrieveTargetRegistryCredentials() {
+    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
+        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
+
     results.targetRegistryCredentials =
         executorService.submit(
             RetrieveRegistryCredentialsStep.forTargetImage(
-                buildConfiguration, childProgressDispatcherFactories.pop()));
+                buildConfiguration, childProgressDispatcherFactory));
   }
 
   private void authenticatePush() {
+    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
+        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
+
     results.pushAuthorization =
         executorService.submit(
             () ->
                 new AuthenticatePushStep(
                         buildConfiguration,
-                        childProgressDispatcherFactories.pop(),
+                        childProgressDispatcherFactory,
                         results.targetRegistryCredentials.get())
                     .call());
   }
 
   private void pullBaseImage() {
+    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
+        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
+
     results.baseImageAndAuth =
         executorService.submit(
-            new PullBaseImageStep(buildConfiguration, childProgressDispatcherFactories.pop()));
+            new PullBaseImageStep(buildConfiguration, childProgressDispatcherFactory));
   }
 
   private void pullAndCacheBaseImageLayers() {
+    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
+        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
+
     results.baseImageLayers =
         executorService.submit(
             () ->
                 scheduleCallables(
                     PullAndCacheBaseImageLayersStep.makeList(
                         buildConfiguration,
-                        childProgressDispatcherFactories.pop(),
+                        childProgressDispatcherFactory,
                         results.baseImageAndAuth.get())));
   }
 
   private void pushBaseImageLayers() {
+    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
+        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
+
     results.baseImageLayerPushResults =
         executorService.submit(
             () ->
                 scheduleCallables(
                     PushLayerStep.makeList(
                         buildConfiguration,
-                        childProgressDispatcherFactories.pop(),
+                        childProgressDispatcherFactory,
                         results.pushAuthorization.get(),
                         results.baseImageLayers.get())));
   }
 
   private void buildAndCacheApplicationLayers() {
+    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
+        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
+
     results.applicationLayers =
         scheduleCallables(
             BuildAndCacheApplicationLayerStep.makeList(
-                buildConfiguration, childProgressDispatcherFactories.pop()));
+                buildConfiguration, childProgressDispatcherFactory));
   }
 
   private void buildImage() {
+    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
+        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
+
     results.builtImage =
         executorService.submit(
             () ->
                 new BuildImageStep(
                         buildConfiguration,
-                        childProgressDispatcherFactories.pop(),
+                        childProgressDispatcherFactory,
                         results.baseImageAndAuth.get().getImage(),
                         realizeFutures(results.baseImageLayers.get()),
                         realizeFutures(Verify.verifyNotNull(results.applicationLayers)))
@@ -180,30 +198,39 @@ public class StepsRunner {
   }
 
   private void pushContainerConfiguration() {
+    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
+        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
+
     results.containerConfigurationPushResult =
         executorService.submit(
             () ->
                 new PushContainerConfigurationStep(
                         buildConfiguration,
-                        childProgressDispatcherFactories.pop(),
+                        childProgressDispatcherFactory,
                         results.pushAuthorization.get(),
                         results.builtImage.get())
                     .call());
   }
 
   private void pushApplicationLayers() {
+    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
+        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
+
     results.applicationLayerPushResults =
         executorService.submit(
             () ->
                 scheduleCallables(
                     PushLayerStep.makeList(
                         buildConfiguration,
-                        childProgressDispatcherFactories.pop(),
+                        childProgressDispatcherFactory,
                         results.pushAuthorization.get(),
                         Verify.verifyNotNull(results.applicationLayers))));
   }
 
   private void pushImage() {
+    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
+        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
+
     results.buildResult =
         executorService.submit(
             () -> {
@@ -213,7 +240,7 @@ public class StepsRunner {
               return new PushImageStep(
                       executorService,
                       buildConfiguration,
-                      childProgressDispatcherFactories.pop(),
+                      childProgressDispatcherFactory,
                       results.pushAuthorization.get(),
                       results.containerConfigurationPushResult.get(),
                       results.builtImage.get())
@@ -222,24 +249,30 @@ public class StepsRunner {
   }
 
   private void loadDocker(DockerClient dockerClient) {
+    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
+        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
+
     results.buildResult =
         executorService.submit(
             () ->
                 new LoadDockerStep(
                         buildConfiguration,
-                        childProgressDispatcherFactories.pop(),
+                        childProgressDispatcherFactory,
                         dockerClient,
                         results.builtImage.get())
                     .call());
   }
 
   private void writeTarFile(Path outputPath) {
+    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
+        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
+
     results.buildResult =
         executorService.submit(
             () ->
                 new WriteTarFileStep(
                         buildConfiguration,
-                        childProgressDispatcherFactories.pop(),
+                        childProgressDispatcherFactory,
                         outputPath,
                         results.builtImage.get())
                     .call());
@@ -251,9 +284,7 @@ public class StepsRunner {
     try (ProgressEventDispatcher progressEventDispatcher =
         ProgressEventDispatcher.newRoot(
             buildConfiguration.getEventHandlers(), rootProgressDescription, stepsToRun.size())) {
-      for (int i = 0; i < stepsToRun.size(); i++) {
-        childProgressDispatcherFactories.push(progressEventDispatcher.newChildProducer());
-      }
+      rootProgressDispatcher = progressEventDispatcher;
 
       stepsToRun.forEach(Runnable::run);
       return results.buildResult.get();

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java
@@ -32,13 +32,14 @@ import com.google.common.util.concurrent.Futures;
 import com.google.common.util.concurrent.ListeningExecutorService;
 import com.google.common.util.concurrent.MoreExecutors;
 import java.nio.file.Path;
+import java.util.ArrayDeque;
 import java.util.ArrayList;
+import java.util.Deque;
 import java.util.List;
 import java.util.concurrent.Callable;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Future;
-import java.util.function.Supplier;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
 
@@ -102,10 +103,8 @@ public class StepsRunner {
   private final List<Runnable> stepsToRun = new ArrayList<>();
 
   @Nullable private String rootProgressDescription;
-  private Supplier<ProgressEventDispatcher.Factory> childProgressDispatcherFactorySupplier =
-      () -> {
-        throw new IllegalStateException("root progress dispatcher uninstantiated");
-      };
+  private Deque<ProgressEventDispatcher.Factory> childProgressDispatcherFactories =
+      new ArrayDeque<>();
 
   private StepsRunner(
       ListeningExecutorService executorService, BuildConfiguration buildConfiguration) {
@@ -117,7 +116,7 @@ public class StepsRunner {
     results.targetRegistryCredentials =
         executorService.submit(
             RetrieveRegistryCredentialsStep.forTargetImage(
-                buildConfiguration, childProgressDispatcherFactorySupplier.get()));
+                buildConfiguration, childProgressDispatcherFactories.pop()));
   }
 
   private void authenticatePush() {
@@ -126,7 +125,7 @@ public class StepsRunner {
             () ->
                 new AuthenticatePushStep(
                         buildConfiguration,
-                        childProgressDispatcherFactorySupplier.get(),
+                        childProgressDispatcherFactories.pop(),
                         results.targetRegistryCredentials.get())
                     .call());
   }
@@ -134,8 +133,7 @@ public class StepsRunner {
   private void pullBaseImage() {
     results.baseImageAndAuth =
         executorService.submit(
-            new PullBaseImageStep(
-                buildConfiguration, childProgressDispatcherFactorySupplier.get()));
+            new PullBaseImageStep(buildConfiguration, childProgressDispatcherFactories.pop()));
   }
 
   private void pullAndCacheBaseImageLayers() {
@@ -145,7 +143,7 @@ public class StepsRunner {
                 scheduleCallables(
                     PullAndCacheBaseImageLayersStep.makeList(
                         buildConfiguration,
-                        childProgressDispatcherFactorySupplier.get(),
+                        childProgressDispatcherFactories.pop(),
                         results.baseImageAndAuth.get())));
   }
 
@@ -156,7 +154,7 @@ public class StepsRunner {
                 scheduleCallables(
                     PushLayerStep.makeList(
                         buildConfiguration,
-                        childProgressDispatcherFactorySupplier.get(),
+                        childProgressDispatcherFactories.pop(),
                         results.pushAuthorization.get(),
                         results.baseImageLayers.get())));
   }
@@ -165,7 +163,7 @@ public class StepsRunner {
     results.applicationLayers =
         scheduleCallables(
             BuildAndCacheApplicationLayerStep.makeList(
-                buildConfiguration, childProgressDispatcherFactorySupplier.get()));
+                buildConfiguration, childProgressDispatcherFactories.pop()));
   }
 
   private void buildImage() {
@@ -174,7 +172,7 @@ public class StepsRunner {
             () ->
                 new BuildImageStep(
                         buildConfiguration,
-                        childProgressDispatcherFactorySupplier.get(),
+                        childProgressDispatcherFactories.pop(),
                         results.baseImageAndAuth.get().getImage(),
                         realizeFutures(results.baseImageLayers.get()),
                         realizeFutures(Verify.verifyNotNull(results.applicationLayers)))
@@ -187,7 +185,7 @@ public class StepsRunner {
             () ->
                 new PushContainerConfigurationStep(
                         buildConfiguration,
-                        childProgressDispatcherFactorySupplier.get(),
+                        childProgressDispatcherFactories.pop(),
                         results.pushAuthorization.get(),
                         results.builtImage.get())
                     .call());
@@ -200,7 +198,7 @@ public class StepsRunner {
                 scheduleCallables(
                     PushLayerStep.makeList(
                         buildConfiguration,
-                        childProgressDispatcherFactorySupplier.get(),
+                        childProgressDispatcherFactories.pop(),
                         results.pushAuthorization.get(),
                         Verify.verifyNotNull(results.applicationLayers))));
   }
@@ -215,7 +213,7 @@ public class StepsRunner {
               return new PushImageStep(
                       executorService,
                       buildConfiguration,
-                      childProgressDispatcherFactorySupplier.get(),
+                      childProgressDispatcherFactories.pop(),
                       results.pushAuthorization.get(),
                       results.containerConfigurationPushResult.get(),
                       results.builtImage.get())
@@ -229,7 +227,7 @@ public class StepsRunner {
             () ->
                 new LoadDockerStep(
                         buildConfiguration,
-                        childProgressDispatcherFactorySupplier.get(),
+                        childProgressDispatcherFactories.pop(),
                         dockerClient,
                         results.builtImage.get())
                     .call());
@@ -241,7 +239,7 @@ public class StepsRunner {
             () ->
                 new WriteTarFileStep(
                         buildConfiguration,
-                        childProgressDispatcherFactorySupplier.get(),
+                        childProgressDispatcherFactories.pop(),
                         outputPath,
                         results.builtImage.get())
                     .call());
@@ -253,7 +251,10 @@ public class StepsRunner {
     try (ProgressEventDispatcher progressEventDispatcher =
         ProgressEventDispatcher.newRoot(
             buildConfiguration.getEventHandlers(), rootProgressDescription, stepsToRun.size())) {
-      childProgressDispatcherFactorySupplier = progressEventDispatcher::newChildProducer;
+      for (int i = 0; i < stepsToRun.size(); i++) {
+        childProgressDispatcherFactories.push(progressEventDispatcher.newChildProducer());
+      }
+
       stepsToRun.forEach(Runnable::run);
       return results.buildResult.get();
 

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java
@@ -103,7 +103,7 @@ public class StepsRunner {
   private final List<Runnable> stepsToRun = new ArrayList<>();
 
   @Nullable private String rootProgressDescription;
-  private Deque<ProgressEventDispatcher.Factory> childProgressDispatcherFactories =
+  private final Deque<ProgressEventDispatcher.Factory> childProgressDispatcherFactories =
       new ArrayDeque<>();
 
   private StepsRunner(

